### ISSUE TITLE
Gluster Server compatibility regression with xattr

### ISSUE THREAD

**Issue by darkain** [openzfs/zfs#16828]
### System information
Type | Version/Name
 --- | ---
Distribution Name	| FreeBSD
Distribution Version	| 14.1-RELEASE
Architecture	| AMD64
OpenZFS Version	| zfs-2.2.4-FreeBSD_g256659204

### Describe the problem you're observing

https://github.com/openzfs/zfs/commit/5c0061345b824eebe7a6578528f873ffcaae1cdd

This commit broke compatibility with the Gluster server application. Gluster relies heavily upon xattrs, more specifically, the "trusted" namespace xattrs. Without these, Gluster server fails to function at all. 

https://github.com/gluster/glusterfs/issues/4340

This issue has been open for months over on the Gluster GitHub repo, however I believe it should also be noted here as a regression in ZFS.

### Describe how to reproduce the problem
A quick repro example without installing Gluster:
```
setextattr user trusted.test test /zfs/test
```
Alternatively: Installer Gluster and attempt to create a volume. `gluster volume create` will fail with the following error message:

### Include any warning/errors/backtraces from the system logs
```
setextattr: /zfs/test: failed: Invalid argument
```
NOTE: Gluster produces a similar error with a different path.

### Testing
https://github.com/openzfs/zfs/blob/f29dcc25c7a784bbb3d3e92942a0a180671a04ac/include/sys/fs/zfs.h#L1956

On my personal machine, I removed the single `ZFS_XA_NS_PREFIX_MATCH(LINUX_TRUSTED, name)` line from `ZFS_XA_NS_PREFIX_FORBIDDEN`, recompiled, and Gluster Server started to function as normal again.

### Upsteam Docs
https://man7.org/linux/man-pages/man7/xattr.7.html
```
       Trusted extended attributes are visible and accessible only to
       processes that have the CAP_SYS_ADMIN capability.  Attributes in
       this class are used to implement mechanisms in user space (i.e.,
       outside the kernel) which keep information in extended attributes
       to which ordinary processes should not have access.
```
This directly states that it is intended to be used by trusted user-space applications, which Gluster Server most likely would fall into. I'm uncertain why ZFS would want to remove this functionality entirely. I understand wanting to unify feature sets between Linux, FreeBSD, and others, but to flat out "forbid" flags like this which is actively used seems to be the wrong decision. At worst, this should be gated with a dataset attribute of some kind, rather than a hard-coded "FORBIDDEN"

**Comment by amotin** [comment]
@darkain If you read FreeBSD's extattr(2) and extattr(9) man pages, you will see that respective extended attribute APIs are taking namespace specification as separate parameter.  What `ZFS_XA_NS_PREFIX_FORBIDDEN()` macro does is preventing namespace specification as part of the name, which is not practiced on FreeBSD and would cause potential (security) problems would the pool be moved between operating systems or may be ion other cases.  There are two namespaces defined now on FreeBSD EXTATTR_NAMESPACE_SYSTEM and EXTATTR_NAMESPACE_USER.  Man page allow file systems to define additional namespaces, but as I can see, there is not trusted namespace currently.  You are welcome to start implementing it from FreeBSD side, if you need it.

**Comment by snajpa** [comment]
This should probably be fixed in the OS+documentation then, it confuses people - there seem to exist setups with OpenZFS 2.2 + FreeBSD + GlusterFS, which rely on the ability to store "trusted" xattrs - I'd wager that we're kinda responsible for this state of things, so it should probably be us to do something about it

[take a look at this StackExchange post](https://unix.stackexchange.com/a/750317)

Now, the FreeBSD [extattr(9)](https://man.freebsd.org/cgi/man.cgi?extattr(9)) manpage says:

> [...] The following
       two namespaces are defined universally, although	individual  file  sys-
       tems  may implement additional namespaces, or not implement these name-
       spaces: EXTATTR_NAMESPACE_USER, EXTATTR_NAMESPACE_SYSTEM.
       
Which, from the point of view of OpenZFS 2.2 users, we did.

So I'd say it's not exactly clear it's OK to say _they_ should change their implementation - we are somewhat responsible for this, aren't we :) It would make sense to me to allow for seamless OS transition between Linux and FreeBSD when it comes to Gluster (and other trusted xattrs users) going forward.

**Comment by snajpa** [comment]
@amotin I'm not familiar with FreeBSD kernel source, but that ^ StackExchange post would suggest it's up to the FS to implement xattrs support more or less completely - as I said, I don't know FreeBSD code, so am taking a leap from that tmpfs note in that post. It's cetainly that way on Linux side - it's upto the FS to uphold the guarantee that only CAP_SYS_ADMIN can access trusted xattrs - so if previously FreeBSD OpenZFS implemented trusted namespace with this guarantee by sharing that logic with the Linux side, I think that was the right, even if trusted xattrs aren't standardized in FreeBSD. It's actually a compatibility design win it could (and used to) work and it didn't require any kernel change, isn't it?

Trusted xattrs are for userspace use, but unlike user.*, regular users just get no access at all.

Gluster uses this to store its own state it needs for managing replication, etc. - in the early days of our vpsAdminOS we didn't support trusted xattrs in our containers and I got Gluster to work by just `s/user\./trusted\./g` on the whole codebase, but that was in the Gluster 2.x days and once that stopped working b/c their code got more complicated, I just went and extended what OpenZFS used to do in a dedicated function to allow first-level user namespace's root users to manage their trusted attrs, cross-container safety guarantee provided by mount namespace. A bit off topic, but demonstrates how it's [up the FS](https://github.com/vpsfreecz/zfs/blob/vpsadminos-release-next/module/os/linux/zfs/zpl_xattr.c#L816) implementation to uphold the trusted xattrs semantics.

**Comment by amotin** [comment]
@snajpa You are missing my focus. I am not arguing about usefulness of "trusted" attributes implementation, I just don't know enough. I am talking that FreeBSD APIs expect namespace to be passed as separate parameter. It should not be sent as part of the name.  So the hack you've done before on FreeBSD should actually look as `s/trusted\.//g` with addition of `EXTATTR_NAMESPACE_USER` argument or `EXTATTR_NAMESPACE_TRUSTED` once it is implemented.

**Comment by snajpa** [comment]
> @snajpa You are missing my focus.

Ah I see, so the problem is what number to use for it and how to guarantee that won't have to change in the future, right? Then I'd be interested in how the old code did it and if there was any discussion around that. I'll take a look tomorrow, out of curiosity

**Comment by snajpa** [comment]
@amotin I just went through current git tip of GlusterFS and it would seem that it uses `EXTATTR_NAMESPACE_USER` on FreeBSD... so I don't think there's any action needed from us, if it doesn't work on current FreeBSD/OpenZFS combo, then it most likely doesn't work with UFS either, which it probably should, at least from the quick glance at the Gluster+kernel code

**Comment by snajpa** [comment]
@darkain if you'd be so kind and could get a syscall trace from glusterfsd when this failure happens so we could confirm it doesn't reach for the trusted namespace, that'd be awesome - I think it doesn't and if there's a bug it's probbably unrelated to the lack of support for trusted xattrs namespace on FreeBSD

**Comment by amotin** [comment]
@snajpa But did they remove "trusted." prefix from the attribute name or they are trying to specify two namespaces in two ways?

**Comment by snajpa** [comment]
@amotin they have these [OS-dependent ifdefs](https://github.com/gluster/glusterfs/blob/7663679e0b2d69ea954d59a42500f5beed001b3e/libglusterfs/src/syscall.c#L550) which I think imply that there was never any interop between Gluster on Linux and Gluster on FreeBSD

**Comment by snajpa** [comment]
umm maybe we're just trying to be smart in the code and actually disallow the combination of `EXTATTR_NAMESPACE_USER` + `trusted.*` in the xattr name?

**Comment by amotin** [comment]
@snajpa Yes. I think so.  And it seems Gluster uses all sort of prefixes:
```
#define GF_NAMESPACE_KEY "trusted.glusterfs.namespace"

#define GF_XATTR_LINKINFO_KEY "trusted.distribute.linkinfo"
#define GFID_XATTR_KEY "trusted.gfid"
#define PGFID_XATTR_KEY_PREFIX "trusted.pgfid."
#define GFID2PATH_VIRT_XATTR_KEY "glusterfs.gfidtopath"
#define GFID2PATH_XATTR_KEY_PREFIX "trusted.gfid2path."
#define GFID2PATH_XATTR_KEY_PREFIX_LENGTH 18
#define VIRTUAL_GFID_XATTR_KEY_STR "glusterfs.gfid.string"
#define VIRTUAL_GFID_XATTR_KEY "glusterfs.gfid"
#define GF_XATTR_MDATA_KEY "trusted.glusterfs.mdata"
```

**Comment by snajpa** [comment]
@amotin yes those virtual ones `glusterfs.*` seem to be used only during run-time, they're filtered out and not saved to disk

**Comment by snajpa** [comment]
olol I should have just read through the commit OP linked :D it's all in there - and now that we know that Gluster would like to store xattrs starting with 'user.*' it's obvious where the problem is. Not quite obvious what to do about it, hm.

**Comment by snajpa** [comment]
Maybe it could be as simple as this?

```diff
diff --git a/include/sys/fs/zfs.h b/include/sys/fs/zfs.h
index 1676020d0..ac01b3ad2 100644
--- a/include/sys/fs/zfs.h
+++ b/include/sys/fs/zfs.h
@@ -1942,8 +1942,6 @@ enum zio_encrypt {
 #define        ZFS_XA_NS_LINUX_SYSTEM_PREFIX_LEN       strlen("system.")
 #define        ZFS_XA_NS_LINUX_TRUSTED_PREFIX          "trusted."
 #define        ZFS_XA_NS_LINUX_TRUSTED_PREFIX_LEN      strlen("trusted.")
-#define        ZFS_XA_NS_LINUX_USER_PREFIX             "user."
-#define        ZFS_XA_NS_LINUX_USER_PREFIX_LEN         strlen("user.")
 
 #define        ZFS_XA_NS_PREFIX_MATCH(ns, name) \
        (strncmp(name, ZFS_XA_NS_##ns##_PREFIX, \
@@ -1953,8 +1951,7 @@ enum zio_encrypt {
        (ZFS_XA_NS_PREFIX_MATCH(FREEBSD, name) || \
            ZFS_XA_NS_PREFIX_MATCH(LINUX_SECURITY, name) || \
            ZFS_XA_NS_PREFIX_MATCH(LINUX_SYSTEM, name) || \
-           ZFS_XA_NS_PREFIX_MATCH(LINUX_TRUSTED, name) || \
-           ZFS_XA_NS_PREFIX_MATCH(LINUX_USER, name))
+           ZFS_XA_NS_PREFIX_MATCH(LINUX_TRUSTED, name)
 
 #ifdef __cplusplus
 }
```

**Comment by snajpa** [comment]
@amotin so my idea is to instead of forbidding certain prefixes entirely on FreeBSD, move it always into the `user.*` xattr namespace - so that `trusted.whatever` would end up stored as `user.trusted.whatever` and would be retrieved when the user asked for `trusted.whatever`. At least when we're on FreeBSD and not in `zfs_xattr_compat` mode - [here's the PoC](a28367c4827083d94bb41ffd72e5c8d1ea3736f7) - I don't think I can do that for Linux but I also don't think similar situation could arise.

There's also the possibility of also allowing `trusted.*` on FreeBSD to be stored and enforced the Linux way, if user on FreeBSD requests `trusted.*` with `EXTATTR_NAMESPACE_USER` - we could store it as 'trusted.*' instead of 'user.*' and enforce the root user semantics.

**Comment by amotin** [comment]
@snajpa The whole goal of that project was to allow compatibility between different OS'es. Adding duplicate prefixes aside if being a mess would not allow that, but instead would make already existing mess even worse.  What Gluster does on FreeBSD I find just broken, and instead of conforming to it we might provide them a way to do it properly.  That might include addition of trusted attributes support to FreeBSD (I have no idea why it is not there, may have a reason or just be an omission), and patching the code in [OS-dependent ifdefs](https://github.com/gluster/glusterfs/blob/7663679e0b2d69ea954d59a42500f5beed001b3e/libglusterfs/src/syscall.c#L550) to cut the prefix from the attribute name and passing it as via the namespace parameter instead of hardcoding `EXTATTR_NAMESPACE_USER`.

**Comment by snajpa** [comment]
@amotin if compatibility between OSes is the primary goal then `trusted.*` in `EXTATTR_NAMESPACE_USER` on FreeBSD could be stored and enforced the Linux way entirely from OpenZFS - and then `EXTATTR_NAMESPACE_TRUSTED` could be added and even that could be supported, on disk both of these variants end up `trusted.*`, there is no number on disk, is there :)

**Comment by snajpa** [comment]
Gluster evidently doesn't really care if the CAP_SYS_ADMIN/root condition is enforced, but I think we have no business telling users in general they can't name the xattrs `user.whatever`, `trusted.whatever` if that's what they'd like - even on FreeBSD that is allowed, isn't it?

**Comment by amotin** [comment]
@snajpa With compatibility I mean on disk. APIs are just different between the OS'es, period. You should not try to push things where they should not go.  Yes, on-disk format for the attributes will likely include the prefixes.  But we can not say that "we have no business telling users in general they can't name the xattrs user.whatever, `trusted.whatever` if that's what they'd like", since nest step they try to read those attributes from user namespace and we would have to allow that, even though process that written them expected "trusted." to have limited access, unless we double-prefix them, which is not what we want either.

**Comment by snajpa** [comment]
https://github.com/openzfs/zfs/issues/16828#issuecomment-2518759813 then this is my best answer, compatible Linux<=>FreeBSD - with that "trusted.*" in EXTATTR_NAMESPACE_USER + root enforced part being optional.

No idea about Illumos and if anyone would care to enforce the root boundary there - but it probably should be, if things shouldn't get more messy. This would IMHO decrease entropy on the user side, but it does increase code complexity, since we have to deal with the `zfs_xattr_compat` non-prefixed way of storing things.

**Comment by snajpa** [comment]
There is the thing that if people now on-disk have `trusted.stuff` and now we only change the code on FreeBSD with `EXTATTR_NAMESPACE_USER` to be prefixed with `user.` in all cases, on disk that translates to `user.trusted.stuff` which is logical, but existing Gluster/FreeBSD installations will have to migrate somehow and we won't have the interop we had before https://github.com/openzfs/zfs/commit/5c0061345b824eebe7a6578528f873ffcaae1cdd

I can already do `user.trusted.stuff` on Linux/ZFS on disk but Gluster doesn't expect that... I have no idea how do they enforce their Linux security model - if they even do - on other platforms, seems like they might not care.

But currently, in master, even if we ignore Gluster, I won't be able to change ondisk stored `user.trusted.stuff` from FreeBSD b/c it's forbidden, probably should be able to read it, from my understanding of the code, while normally able to both change and read `user.untrusted.stuff` (see the paradox?)

```
root@fbsd-14-1:~ # setextattr user untrusted.test lolo f
root@fbsd-14-1:~ # setextattr user trusted.test lolo f
setextattr: f: failed: Invalid argument
```

This code would be a whole lot easier if we could rely that there's only the new prefixed way of storing xattrs.

**Comment by amotin** [comment]
> now we only change the code on FreeBSD with EXTATTR_NAMESPACE_USER to be prefixed with user. in all cases, on disk that translates to user.trusted.stuff

That is why I'd prefer adding the proper namespace to FreeBSD and patch the Gluster. It would avoid double-prefixing and zillion other problems.  Code before https://github.com/openzfs/zfs/commit/5c0061345b824eebe7a6578528f873ffcaae1cdd added no prefixes so should migrate to addition of EXTATTR_NAMESPACE_TRUSTED just fine.  Just Gluster would need to be updated together with OS.

**Comment by snajpa** [comment]
@amotin but then `user.trusted.stuff` will be the result of OpenZFS+Gluster+Illumos with `zfs_xattr_compat=0` if that even exists (have no idea what Illumos ZFS code looks like these days)

**Comment by snajpa** [comment]
It'd be awesome if there was a way how out of this by upgrading within supported release versions of FreeBSD/illumos + OpenZFS with the trusted semantics respected with a non-hacky implementation.

@allanjude my first project? :D

**Comment by amotin** [comment]
@snajpa I would agree that `trusted.test` may be a valid user namespace xattr if it means that on Linux it will be accessed as `user.trusted.test`.  But it is not true for Gluster, and I bet most of software that may appear doing the same.  The way Gluster does it is incorrect and not cross-platform compatible!  We should not stimulate improper behavior!  We should provide a right way, and may be, if it does not create more problems, turn a blind eye on others.  Broken software can be fixed once and the problem will be forgotten in few years, while supporting improper behavior will extend it into eternity.

I feel like I am repeating myself third time and you don't hear me.  I've already told how I see the proper final solution.  And if on top of that you can propose temporary transitional shims -- I'll give them a fair review.  But I will not accept affirmation of improper API usage, considering the consequences.

**Comment by snajpa** [comment]
@amotin yes it would be accessed as `user.trusted.test` on Linux. I do hear you, I'm just saying that currently we're in a business of telling FreeBSD users what they can't store `EXTATTR_NAMESPACE_USER` in terms of keys, the core point is that this shouldn't have happened in the first place. I do recognize what you're saying, I just think that we should change this primarily - as this is our bug.

Gluster's current improper usage of xattr APIs on top of OpenZFS across supported platform will need to be addressed more broadly than just releasing a fixed version that uses new untrusted xattr namespace in FreeBSD - because people's requirements are different, you can see that a lot of times in issues here too - some have policies that they can upgrade this or that component, but that one closer to the core of their business needs to undergo more rigorous testing and sometimes even certifications, etc., so I can imagine there could be situations where people are unwilling/unable to upgrade Gluster itself in lockstep with kernel+ZFS. Others might be able to upgrade Gluster but kernel+ZFS upgrade could be late - though I think we're okay in this particular scenario. For the former, some script for one-off migration between improper xattrs in `user.trusted.*` to `trusted.*` seems like it could address the issue. I don't think we can avoid having a solution here (by we I mean Gluster+OpenZFS teams together, not just us).

**Comment by amotin** [comment]
@snajpa I am not sure we have the `user.trusted.*` situation now.  IIRC before this change ZFS on FreeBSD was not adding `user.` prefixes.  After the change it returns error instead.  Removal of the condition now would actually get us the problem, requiring some mentioned fixup later.  Gluster could have `#ifdef EXTATTR_NAMESPACE_TRUSTED` to switch between the behaviors, but without some additional efforts it would have to be updated in lock-step with kernel+ZFS to work.  Since most FreeBSD users likely use in-base ZFS version, having kernel to ZFS sync is not a big problem.  Remains question of sync with Gluster.  At worst we could require its rebuild (if the ifdef is planted there in advance) with new major FreeBSD version, but yea some nicer transition plan would be nice to have, unless it creates more problems than solves.

**Comment by snajpa** [comment]
If we can accept that we'll only fix this slowly, if we say that it's ok for some parts of this to stay broken for a while - then we can use time to our advantage -

- introduce trusted xattrs ns in FreeBSD
- enable support for them in OpenZFS
- wait till most FreeBSD/Gluster deployments are up to date
- only then fix the problem with `user.trusted.*`, `user.user.*`, `user.system.*` being unreachable with -EINVAL on FreeBSD

is that really okay though? it's a bet on the theory there's no other significant breakage in other software caused by the `-EINVAL`s :)

**Comment by amotin** [comment]
@snajpa "only then fix the problem with user.trusted.*, user.user.*, user.system.* being unreachable with -EINVAL on FreeBSD" -- I am still not sure it is a big problem, but yes.  Though I would consider the problem solved after the step 2.  We could add a tunable to allow disabling the check with a big warning that xattrs written that way might be inaccessible on different platforms or when the possibly broken software is finally fixed.

**Comment by kohlschuetter** [comment]
Just adding some perspective:

What @snajpa said,
> [@amotin](https://github.com/amotin) yes those virtual ones `glusterfs.*` seem to be used only during run-time, they're filtered out and not saved to disk

may not be what I see here on Linux (with OpenZFS).

On my zfs-backed gluster bricks, I see the following attributes:

```
afr.dirty
gfid
gfid2path.0b0302ac59fb4d27
gfid2path.0f5f7ddfed6dcbea
gfid2path...(a lot more like these)
gfid2path.fbc96ef7f492ccb3
glusterfs.dht
glusterfs.dht.linkto
glusterfs.mdata
glusterfs.volume-id
```

I didn't see any `trusted.` attributes, but I assume they only appear when the device is in some specific state (e.g. `trusted.ec.dirty`, `trusted.io-stats-dump`)

In Linux, I can access these with `attr -R -g ATTRNAME FILENAME`, i.e. specifying the "root" attribute namespace (`attr` is a tool that apparently originated from XFS but seems to work on all kinds of filesystems on Linux)

Without looking into what glusterfs does, whether it does it the right way or not, it would be helpful to know what happens when someone on FreeBSD mounts an ZFS pool/FS that was generated on Linux containing such attributes.

You can easily reproduce this without even knowing what glusterfs is or does:

Set attribute:
```
attr -R -s trusted.ec.dirty.testing -V my-test-value SOME-FILE
```

Verify attribute (on Linux)
```
# show value
attr -R -g trusted.ec.dirty.testing SOME-FILE
# doesn't work, because "root" namespace was not specified
attr -g trusted.ec.dirty.testing SOME-FILE
```

Then verify on FreeBSD with lsextattr, getextattr, etc.
(And maybe try the opposite direction, too)

I understand that the "root" namespace on Linux should map to the "user" namespace on FreeBSD.

I also understand that somehow there both exists a concept of "attribute namespace" on Linux (attr -R for "root", attr -S for "security"), but also some kind of by-convention namespace prefixing. I don't know how these are related when it comes down to the bits on-disk, though.

My understanding is that having such attributes on a ZFS pool should yield the same results on Linux and FreeBSD, and maybe filtering anything by name (i.e.., prefix "trusted.") is probably not necessary. — I don't think we need to enforce some Linux-specific security logic when mounting on FreeBSD.

What would also be interesting of course is how Linuxulator handles that...

**Comment by erwinlem** [comment]
I have some patches adding a gluster prefix on write and removing it on read. This should make it possible to mix bricks of different oses. Swapping os and using the same physical disk is not possible, you can't have it all. Prefixing all the attributes might not be a bad thing in general.

https://github.com/erwinlem/Freebsd-CustomPorts/tree/master/filesystems/glusterfs
